clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -603,19 +586,24 @@ public synchronized boolean hasRequest() {
                     } else
                         break;
                 }
-                if (requestEntry.getValue().isEmpty())
+                if (requests.isEmpty())
                     iterator.remove();
             }
             return expiredRequests;
         }
 
-        public synchronized List<ClientRequest> remove(Node node) {
-            List<ClientRequest> requests = unsent.remove(node);
-            return requests == null ? Collections.<ClientRequest>emptyList() : requests;
+        public synchronized ConcurrentLinkedQueue<ClientRequest> remove(Node node) {
+            ConcurrentLinkedQueue<ClientRequest> requests = unsent.remove(node);
+            return requests == null ? new ConcurrentLinkedQueue<ClientRequest>() : requests;
+        }
+
+        public synchronized Iterator<ClientRequest> requestIterator(Node node) {
+            ConcurrentLinkedQueue<ClientRequest> requests = unsent.get(node);
+            return requests == null ? null : requests.iterator();
```

Comment:
The benefits of having `remove(...)` return empty list instead of null is that `failUnsentRequests(...)` and `checkDisconnects(...)` can be simpler since they don't need to check if the result is null. I think it is fine since `UnsentRequests` is just an private class used only in `ConsumerNetworkClient`. I will check it to allow `remove()` to return null if you prefer that.